tianshou/data/collector.py

```diff
@@ -543,52 +892,129 @@ def _collect(
+                # local_idx - see block comment on class level
+                # Step 7
+                env_done_local_idx_D = np.where(done_R)[0]
+                episode_lens_D = ep_len_R[env_done_local_idx_D]
+                episode_returns_D = ep_return_R[env_done_local_idx_D]
+                episode_start_indices_D = ep_start_idx_R[env_done_local_idx_D]
+
+                episode_lens.extend(episode_lens_D)
+                episode_returns.extend(episode_returns_D)
+                episode_start_indices.extend(episode_start_indices_D)
+
+                # Step 8
                 # now we copy obs_next to obs, but since there might be
                 # finished episodes, we have to reset finished envs first.
-
                 gym_reset_kwargs = gym_reset_kwargs or {}
+                # The index env_done_idx_D was based on 0, ..., R
+                # However, each env has an index in the context of the vectorized env and buffer. So the env 0 being done means
+                # that some env of the corresponding "global" index was done. The mapping between "local" index in
+                # 0,...,R and this global index is maintained by the ready_env_ids_R array.
+                # See the class block comment for more details
+                env_done_global_idx_D = ready_env_ids_R[env_done_local_idx_D]
                 obs_reset_DO, info_reset_D = self.env.reset(
-                    env_id=env_ind_global_D,
+                    env_id=env_done_global_idx_D,
                     **gym_reset_kwargs,
                 )
 
                 # Set the hidden state to zero or None for the envs that reached done
                 # TODO: does it have to be so complicated? We should have a single clear type for hidden_state instead of
                 #  this complex logic
-                self._reset_hidden_state_based_on_type(env_ind_local_D, last_hidden_state_RH)
+                self._reset_hidden_state_based_on_type(env_done_local_idx_D, last_hidden_state_RH)
+
+                # Step 9
+                # execute episode hooks for those envs which emitted 'done'
+                for local_done_idx, cur_ep_return in zip(
+                    env_done_local_idx_D,
+                    episode_returns_D,
+                    strict=True,
+                ):
+                    cur_ep_index_slice = slice(
+                        ep_start_idx_R[local_done_idx],
+                        insertion_idx_R[local_done_idx] + 1,
+                    )
+
+                    (
+                        cur_ep_index_array,
+                        cur_ep_batch,
+                    ) = self._get_buffer_index_and_entries_for_episode_from_slice(
+                        cur_ep_index_slice,
+                    )
+                    cur_ep_batch = cast(EpisodeBatchProtocol, cur_ep_batch)
+
+                    # Step 10
+                    episode_hook_additions = self.run_on_episode_done(cur_ep_batch)
```

Comment:
Maybe more of an aesthetics thing, but the on step hook modifies the batch in place, while the episode done hook returns a dictionary and does the buffer manipulation in sight

Reply:
Yes, it's because the episode hook operates on a batch that was already added to the buffer. It can't modify the buffer directly. The step hook operates before the transition is added, so it can to anything in place. So they need different signatures